Add a Portal component for rendering content outside of the current DOM element (#1025)
primer/components: docs/content/Portal.mdx

---
title: Portal
---

Portals allow you to create a separation between the logical React component hierarchy and the physical DOM. See the [React documentation on portals](https://reactjs.org/docs/portals.html) for an in-depth explanation.

This Portal component will render all children into the portal root DOM node instead of as children of this Portal's parent DOM element. This is useful for breaking out of the current stacking context. For example, popup menus and tool tips may need to render on top of (read: covering up) other UI. The best way to guarantee this is to add these elements to top-level DOM, such as directly on document.body. These elements can then be moved to the correct location using absolute positioning.

## Customizing the portal root

> By default, Primer will create a portal root for you as a child of `document.body`. If you would like to specify your own portal root, there are two options:

Reviewer: Should this be updated to say that it's added by default to the `BaseStyles` component which all applications should have as high up as possible in their component tree?

1. Before rendering a `<Portal>` for the first time, ensure that an element exists with id `__primerPortalRoot__`. If that exists, it will be used as the default portal root.

> 2. Call the `registerPortalRoot` function, passing in the element you would like to use as your default portal root.

Reviewer: We should also add a warning to this section about what can happen if you put your custom portal root above BaseStyles and/or your ThemeProvider

---
title: Portal
---

Portals allow you to create a separation between the logical React component hierarchy and the physical DOM. See the [React documentation on portals](https://reactjs.org/docs/portals.html) for an in-depth explanation.

> This Portal component will render all children into the portal root DOM node instead of as children of this Portal's parent DOM element. This is useful for breaking out of the current stacking context. For example, popup menus and tool tips may need to render on top of (read: covering up) other UI. The best way to guarantee this is to add these elements to top-level DOM, such as directly on document.body. These elements can then be moved to the correct location using absolute positioning.

Reviewer: ```suggestion
This Portal component will render all children into the portal root DOM node instead of as children of this Portal's parent DOM element. This is useful for breaking out of the current stacking context. For example, popup menus and tooltips may need to render on top of (read: covering up) other UI. The best way to guarantee this is to add these elements to top-level DOM, such as directly on `document.body`. These elements can then be moved to the correct location using absolute positioning.
```

## Customizing the portal root

> By default, Primer will create a portal root for you as a child of the closest `<BaseStyles>` element, or `document.body` of none is found. If you would like to specify your own portal root, there are two options:

Reviewer: ```suggestion
By default, Primer will create a portal root for you as a child of the closest `<BaseStyles>` element, or `document.body` if none is found. If you would like to specify your own portal root, there are two options:
```

1. Before rendering a `<Portal>` for the first time, ensure that an element exists with id `__primerPortalRoot__`. If that exists, it will be used as the default portal root.
2. Call the `registerPortalRoot` function, passing in the element you would like to use as your default portal root.

> Keep in mind that any inherited styles applied to portaled elements are based on its physical DOM parent. Practically this means that styles added by a `<BaseStyles>` element will not apply to the portaled content unless the portal root is a descendent of a `<BaseStyles>` element.

Reviewer: ```suggestion
Keep in mind that any inherited styles applied to portaled elements are based on its browser DOM parent. Practically this means that styles added by a `<BaseStyles>` element will not apply to the portaled content unless the portal root is a descendent of a `<BaseStyles>` element.
```
